### DISCUSSION TITLE
Access modify the DOM and fire keyboard and mouse events for the webview

### DISCUSSION THREAD

**Discussion by insinfo** [cefsharp/CefSharp#3975] [General]
I would like to be able to access modify the DOM and fire keyboard and mouse events for the webview to for example click a button and or fill in an input programmatically similar to lib dotnetbrowser https://dotnetbrowser-support.teamdev.com/docs/guides /gs/dom.html#dom

**Comment by amaitland** [comment]
You can use the CefSharp specific port of puppeteer Sharp for this purpose. 

https://github.com/cefsharp/Puppeteer

**Comment by insinfo** [comment]
@amaitland

I haven't seen any documentation on how to modify the DOM and execute keyboard and mouse events in the webview at this link you posted.

But I think it would be more interesting if this feature was already available directly in CefSharp as well as the dotnetbrowser lib.

**Reply by amaitland** [reply]
> I haven't seen any documentation on how to modify the DOM and execute keyboard and mouse events in the webview at this link you posted.

If you look at the unit tests there are lots of examples, here are just a few

- https://github.com/cefsharp/Puppeteer/blob/main/lib/PuppeteerSharp.Tests/KeyboardTests/KeyboardTests.cs
- https://github.com/cefsharp/Puppeteer/blob/main/lib/PuppeteerSharp.Tests/MouseTests/MouseTests.cs
- https://github.com/cefsharp/Puppeteer/tree/main/lib/PuppeteerSharp.Tests/QuerySelectorTests

As it's largely a port of puppeteer Sharp most of the documentation is relevant. 

If you'd like to sponsor my time then I'd be happy to provide you with specific examples. 
https://github.com/sponsors/amaitland

> But I think it would be more interesting if this feature was already available directly in CefSharp as well as the dotnetbrowser lib.

It's technically infeasible to add an API that mirrors dotnetbrowser. CefSharp is just one of many chromium embedded framework wrappers. CEF doesn't have a matching API. 

Maintaining CefSharp puppeteer as a separate project allows for easier syncing with the upstream puppeteer Sharp project.
